### PR TITLE
JEP 18137: Scope of JAX NumPy & SciPy Wrappers

### PR DESCRIPTION
**Pull request by jakevdp**
This is adds a JAX Enhancement Proposal (JEP) discussing the scope of NumPy & SciPy wrappers going forward. Comments welcome!

Rendered preview: https://jax--18137.org.readthedocs.build/en/18137/jep/18137-numpy-scipy-scope.html

### CODE REVIEW

**Comment by chrisflesher** [comment]
First of all I'd like to thank the JAX team for your efforts. You guys are my heros!

Our group is developing underwater mapping software, we have found it great to be able to try radically different ideas just using the standard numpy / scipy API. This was one of the many things that convinced us to try JAX.

I enjoyed contributing to this project with the `Rotation` / `Slerp` classes. Before writing these we were using a 3rd party library,  `jax_transformations3d`. Switching to a common API allowed us to try some radically different ideas because we could refactor 2000+ lines of code to use either scipy, JAX, or numba.

For `Axis 5: Functional vs. Object-Oriented APIs` I'm wondering what the root reasons are for why all classes are being avoided? I agree with you they can be difficult to use / implement properly compared to pure functions, but from a first principles perspective is there a reason why they should be ruled out entirely? For example I thought all functions in the `Rotation` class should be fully compatible with JAX `vmap` / `pmap` calls with arbitrary number of input dimensions (due to the `jax.numpy.vectorize` wrappers)?

For `Axis 1: XLA alignment` we use the `unique` call a lot! (to populate underwater maps). I guess you say in `Axis 2` this is an exception... whew, was worried there for a sec!

I had submitted a PR for the `CubicHermiteSpline` class recently but this seems to violate `Axis 3` and `Axis 5`. I was hoping to use this to implement the algorithm "A General and Adaptive Loss Function" in the `optax` repo eventually (while encouraging code reuse)... maybe not a big deal to ax it short term but may consider if you really want to discourage implementing large swaths of the API long term? For example I would find it convenient to have a "batteries included" implementation of a spline when implementing a camera trajectory.

I had been thinking about implementing some spline or ndimage stuff as a fun hobby project, it seems like that stuff is going to be actively discouraged though :(

**Comment by chrisflesher** [comment]
Well I closed the `CubicHermiteSpline` PR, it sounds like it isn't going to make it. Let me know what you want to do about the `Rotation` and `Slerp` classes, I guess I wouldn't mind maintaining these in a separate repo if that would help anything but might need some help figuring out licensing.

**Comment by chrisflesher** [comment]
Here's what I'm thinking the 3rd party `jax.scipy.spatial` module would look like if you guys decided to go this route, any comments are welcome:
https://github.com/chrisflesher/jax-scipy-spatial/tree/transform

File: docs/jep/18137-numpy-scipy-scope.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+*October 2023*
+
+Until now, the intended scope of {mod}`jax.numpy` and {mod}`jax.scipy` has been relatively
+ill-defined. This document proposes a well-defined scope for these packages to better guide
+and evaluate future contributions, and to motivate the removal of some out-of-scope code.
+
+## Background
+From the beginning, JAX has aimed to provide a NumPy-like API for executing code in XLA,
+and a big part of the project’s development has been building out the {mod}`jax.numpy` and
+{mod}`jax.scipy` namespaces as XLA-backed wrappers of NumPy and SciPy APIs. There has always
```
Comment (froystig) [inline docs/jep/18137-numpy-scipy-scope.md:14]:
```suggestion
{mod}`jax.scipy` namespaces as JAX-based implementations of NumPy and SciPy APIs. There has always
```

... since they support all jax transformations.

File: docs/jep/18137-numpy-scipy-scope.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+ {func}`numpy.unique` (`unique_all`, `unique_counts`, `unique_inverse`, `unique_values`) which
+suggests that, despite the function not being precisely aligned with XLA, it is important
+enough to the Python user community that JAX should perhaps implement it.
+
+### Axis 3: Existence of Downstream Implementations
+For functionality that does not align with Axis 1 or 2, an important consideration for
+inclusion into JAX is whether there exist well-supported downstream packages that supply
+the functionality in question. A good example of this is {mod}`scipy.optimize`: while JAX does
+include a minimal set of wrappers of {mod}`scipy.optimize` functionality, a much more complete
+treatment exists in the [JAXOpt](https://jaxopt.github.io/) package, which is actively
```
Comment (froystig) [inline docs/jep/18137-numpy-scipy-scope.md:68]:
```suggestion
treatment exists in the [JAXopt](https://jaxopt.github.io/) package, which is actively
```

File: docs/jep/18137-numpy-scipy-scope.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+the {mod}`scipy.optimize` API was too constraining, and different teams began working on the
+[JAXopt](https://jaxopt.github.io/) package and the
+[Optimistix](https://github.com/patrick-kidger/optimistix) package, each of which contain
+a much more comprehensive and better-tested set of optimization routines in JAX.
+
+Because of these well-supported external packages, we now consider {mod}`scipy.optimize`
+to be out-of-scope for JAX.
+
+*Recommendation: deprecate {mod}`jax.scipy.optimize` and/or make it a lightweight wrapper
+around an optional JAXOpt or Optimistix dependency.*
```
Comment (froystig) [inline docs/jep/18137-numpy-scipy-scope.md:250]:
```suggestion
around an optional JAXopt or Optimistix dependency.*
```

**Review by froystig — APPROVED** [review]
Nice work!
